From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: zcrypt: fix z90crypt kernel modules causes 100% cpu usage
References: bnc#518291,LTC#54925

Symptom:     100% cpu usage
Problem:     ap_poll_timer expires permanently and is restarts
Solution:    Only restart ap_poll_timer when it expires and do it in the
             future.

Acked-by: John Jolly <jjolly@suse.de>
---
Index: linux-sles11/drivers/s390/crypto/ap_bus.c
===================================================================
--- linux-sles11.orig/drivers/s390/crypto/ap_bus.c
+++ linux-sles11/drivers/s390/crypto/ap_bus.c
@@ -930,10 +930,16 @@ ap_config_timeout(unsigned long ptr)
  */
 static inline void ap_schedule_poll_timer(void)
 {
+	ktime_t hr_time;
 	if (hrtimer_is_queued(&ap_poll_timer))
 		return;
-	hrtimer_start(&ap_poll_timer, ktime_set(0, poll_timeout),
-		      HRTIMER_MODE_ABS);
+
+	if (ktime_to_ns(hrtimer_get_remaining(&ap_poll_timer)) <= 0) {
+		hr_time = ktime_set(0, poll_timeout);
+		hrtimer_forward_now(&ap_poll_timer, hr_time);
+		hrtimer_restart(&ap_poll_timer);
+	}
+	return;
 }
 
 /**
